### PR TITLE
refactor(semantic): rewrite template resolver — eliminate SubstType, fix crashes

### PR DESCRIPTION
**Pull request by 16bit-ykiko**
## Summary

Rewrites the template resolver to eliminate `SubstType`/`CodeSynthesisContexts` dependency, fixing widespread crashes on real-world C++ code.

### What changed

**Architecture**: replaced double-TreeTransform (PseudoInstantiator + SubstType) with single-layer design:
- **`SubstituteOnly`** — new lightweight TreeTransform for Phase 2 (typedef expansion + parameter substitution). Does NOT override `TransformDependentNameType`, breaking the typedef ↔ lookup infinite cycle.
- **`PseudoInstantiator`** — retains heuristic lookup (the unique value clang doesn't provide), delegates substitution to `SubstituteOnly`.

**Deleted**:
- `DesugarOnly` class
- `instantiate()` method and all `CodeSynthesisContexts` / `SubstType` usage
- `state()` / `rewind()` stack management
- `std::abort()` on valid NNS kinds
- `#ifndef NDEBUG` debug flag + `std::print` logging

**Added**:
- `SubstituteOnly` class with depth guard
- `InstantiationStack::findArgument()` — depth/index based parameter lookup
- CTD→TST resolution for `DependentTemplateSpecializationType` (enables `__alloc_traits::rebind<T>::other` resolution)
- `active_resolutions` (DNT cycle detection) + `active_ctd_lookups` (CTD cycle detection via RAII guard)
- Stack frame pollution fix: pop lookup frames before further `TransformType`
- Pack argument support (single-element forwarding)
- Null safety on all Transform return paths
- Structured `LOG_DEBUG` trace logging with indentation
- `--log-level` / `--test-filter` CLI options for unit test runner
- Bounds checks in `hole()`, `ResugarOnly`, `visitTemplateDeclContexts`

**Tests**: 20 → 36 passing test cases (+5 documented TODOs for known limitations). New coverage: recursive base classes, multiple inheritance, typedef chains, CRTP, `remove_reference` partial specs, `std::map`, `std::basic_string`, pack forwarding.

### Stress test result

```
CDB: llvm-project build (4669 C++ files)
Types resolved: 3,690,190
Types unchanged: 75,907,298
Crashes: 0
```

Before this PR, the same test produced ~52% crash rate (413 crashes in 800 files).

### Known limitations (documented as TODOs)

- NTTP partial specialization matching (`enable_if<true, X>`, `A<X, 0>`)
- Template template parameter deduction
- Non-dependent qualifier nested class templates (`Outer<int>::Inner<X>`)
- Multi-element pack expansion
- `CXXDependentScopeMemberExpr` lookup (unimplemented)
- Operator-name lookup in dependent contexts

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Release Notes

* **New Features**
  * Enhanced template resolution with improved cycle detection to prevent infinite loops
  * Better type substitution handling for complex dependent types

* **Bug Fixes**
  * Fixed edge cases in template specialization resolution
  * Improved null-safety in type transformations
  * Enhanced handling of standard library template traits

* **Tests**
  * Expanded test coverage for recursive and complex template patterns
  * Added validation for standard library type resolution

### CODE REVIEW

File: src/semantic/resolver.cpp [hunk 1]
```diff
@@ -191,9 +98,149 @@ struct InstantiationStack {
...
+            }
+            if(params && params->getDepth() == depth) {
+                if(index < it->second.size()) {
+                    return &it->second[index];
+                }
+                return nullptr;
+            }
+        }
+        return nullptr;
+    }
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply src/semantic/resolver.cpp]: Addressed in d1468c9.

File: src/semantic/resolver.cpp [hunk 2]
```diff
@@ -191,9 +98,149 @@ struct InstantiationStack {
...
+
+/// Helper to extract underlying type from a Decl.
+static clang::QualType getDeclType(clang::Decl* decl) {
+    if(!decl)
+        return clang::QualType();
+    if(auto* TND = llvm::dyn_cast<clang::TypedefNameDecl>(decl))
+        return TND->getUnderlyingType();
+    if(auto* RD = llvm::dyn_cast<clang::RecordDecl>(decl))
+        return clang::QualType(RD->getTypeForDecl(), 0);
+    return clang::QualType();
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply src/semantic/resolver.cpp:171]: Addressed in d1468c9.

File: src/semantic/resolver.cpp [hunk 3]
```diff
@@ -341,89 +386,88 @@ class PseudoInstantiator : public clang::TreeTransform<PseudoInstantiator> {
...
+                auto type = getDeclType(decl);
+                if(!type.isNull()) {
+                    type = substitute(type);
+                }
+                while(stack.data.size() > stackSize) {
+                    stack.pop();
+                }
+                if(!type.isNull()) {
+                    resolved.try_emplace(NNS, type);
+                    return lookup(type, name);
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply src/semantic/resolver.cpp:565]: Addressed in d1468c9.

File: src/semantic/resolver.cpp [hunk 4]
```diff
@@ -689,104 +647,210 @@ class PseudoInstantiator : public clang::TreeTransform<PseudoInstantiator> {
...
+                    if(auto ET = llvm::dyn_cast<clang::ElaboratedType>(type)) {
+                        type = ET->getNamedType();
+                    }
+                    TLB.pushTrivial(context, type, {});
+                    return type;
+                }
+            }
+        }
+        return Base::TransformTypedefType(TLB, TL);
     }
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply src/semantic/resolver.cpp]: Addressed in d1468c9.

File: tests/unit/unit_tests.cc [hunk 5]
```diff
@@ -1,26 +1,45 @@
...
+static std::string_view find_arg(int argc, const char** argv, const char* name) {
+    for(int i = 1; i < argc; ++i) {
+        // --name=value
+        auto len = std::strlen(name);
+        if(std::strncmp(argv[i], name, len) == 0 && argv[i][len] == '=') {
+            return argv[i] + len + 1;
+        }
+        // --name value
+        if(std::strcmp(argv[i], name) == 0 && i + 1 < argc) {
+            return argv[i + 1];
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/unit/unit_tests.cc]: Addressed in d1468c9.

File: tests/unit/unit_tests.cc [hunk 6]
```diff
@@ -1,26 +1,45 @@
...
+            clice::logging::options.level = clice::logging::Level::trace;
+        } else if(log_level == "debug") {
+            clice::logging::options.level = clice::logging::Level::debug;
+        } else if(log_level == "info") {
+            clice::logging::options.level = clice::logging::Level::info;
+        } else if(log_level == "warn") {
+            clice::logging::options.level = clice::logging::Level::warn;
+        } else if(log_level == "err") {
+            clice::logging::options.level = clice::logging::Level::err;
+        }
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/unit/unit_tests.cc]: Addressed in d1468c9.

File: tests/unit/unit_tests.cc [hunk 7]
```diff
@@ -1,26 +1,52 @@
...
+            return argv[i] + len + 1;
+        }
+        // --name value (reject if next token looks like another option)
+        if(std::strcmp(argv[i], name) == 0 && i + 1 < argc &&
+           std::strncmp(argv[i + 1], "--", 2) != 0) {
+            return argv[i + 1];
+        }
+    }
+    return {};
+}
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/unit/unit_tests.cc]: Addressed in d1468c9.

**Comment by 16bit-ykiko** [comment]
@coderabbitai review
